Erc721 votes and general Votes component (#1114)
OpenZeppelin/cairo-contracts: docs/modules/ROOT/pages/governance.adoc

            'DAPP_VERSION'
        }
    }

    // We need to call the VotesComponent::transfer_voting_units function 
    // after every mint, burn and transfer.
    // For this, we use the ERC20Component::ERC20HooksTrait.
    impl ERC20VotesHooksImpl<
        TContractState,
        impl Votes: VotesComponent::HasComponent<TContractState>,
        impl HasComponent: ERC20Component::HasComponent<TContractState>,
        +NoncesComponent::HasComponent<TContractState>,
        +Drop<TContractState>
    > of ERC20Component::ERC20HooksTrait<TContractState> {
        fn after_update(
            ref self: ERC20Component::ComponentState<TContractState>,
            from: ContractAddress,
            recipient: ContractAddress,
            amount: u256
        ) {
            let mut votes_component = get_dep_component_mut!(ref self, Votes);
            votes_component.transfer_voting_units(from, recipient, amount);
        }
    }

    #[constructor]
    fn constructor(ref self: ContractState) {
        self.erc20.initializer("MyToken", "MTK");
    }
}
----

The VotesComponent will automatically track voting power as tokens are transferred, minted, or burned.

> For a detailed API reference, see the xref:api/governance.adoc#VotesComponent[VotesComponent API documentation].

Author: Yeah.... I don't think it adds a lot of value. So we can rely on the link at the botton of the page. Added the interface section as well.
```
This is the full interface of the `VotesImpl` implementation:
[source,cairo]
----
#[starknet::interface]
pub trait IVotes<TState> {
    // IVotes
    fn get_votes(self: @TState, account: ContractAddress) -> u256;
    fn get_past_votes(self: @TState, account: ContractAddress, timepoint: u64) -> u256;
    fn get_past_total_supply(self: @TState, timepoint: u64) -> u256;
    fn delegates(self: @TState, account: ContractAddress) -> ContractAddress;
    fn delegate(ref self: TState, delegatee: ContractAddress);
    fn delegate_by_sig(ref self: TState, delegator: ContractAddress, delegatee: ContractAddress, nonce: felt252, expiry: u64, signature: Array<felt252>);

    // INonces
    fn nonces(self: @TState, owner: ContractAddress) -> felt252;
}
----
```

Based on other examples we try to include the full abi that the component generates, I have a few quesitons / topics for disucssion
- Is it accurate to say this is the interface of the implementation?
- I saw in many components we have an `IComponent`interface and another `ComponentABI`interface, which has more functions related to other components. I'm not sure I understand the reason to having both and if we should take into account here in the docs and in the component overall

> For a detailed API reference, see the xref:api/governance.adoc#VotesComponent[VotesComponent API documentation].

Reviewer: The name of this trait should be `VotesABI` and not `IVotes`, since it contains INonces, and we are presenting ABIs in other interface sections. We should also add VotesABI to the interface file under the votes directory, since now we have the component.

ABIs are traits containing all the external functions that must/can be implemented when a component is in use. If SRC5 is a requirement for a component (like AcessControlComponent), then supports_interface should be part of the ABI even when it is not defined in the component itself. When there are optional external functions that may or may not be embedded when using the component, those should be included in the ABI as well. The idea of ABIs is to act as a single dispatcher to interact with the component without having to define multiple different ones.

> For a detailed API reference, see the xref:api/governance.adoc#VotesComponent[VotesComponent API documentation].

Author: Ok, thanks for the explanation. Went ahead with this for the ABI:

```
/// Common interface to interact with the `Votes` component.
#[starknet::interface]
pub trait VotesABI<TState> {
    // Votes
    fn get_votes(self: @TState, account: ContractAddress) -> u256;
    fn get_past_votes(self: @TState, account: ContractAddress, timepoint: u64) -> u256;
    fn get_past_total_supply(self: @TState, timepoint: u64) -> u256;
    fn delegates(self: @TState, account: ContractAddress) -> ContractAddress;
    fn delegate(ref self: TState, delegatee: ContractAddress);
    fn delegate_by_sig(
        ref self: TState,
        delegator: ContractAddress,
        delegatee: ContractAddress,
        nonce: felt252,
        expiry: u64,
        signature: Array<felt252>
    );

    // Nonces
    fn nonces(self: @TState, owner: ContractAddress) -> felt252;
}
```

> When there are optional external functions that may or may not be embedded when using the component, those should be included in the ABI as well.
In this case ERC20 and ERC721 functions(and their depenedencies) may be part of the public ABI of someone using the VotesComponent, but I imagine we don't want to include all of this here right?

> For a detailed API reference, see the xref:api/governance.adoc#VotesComponent[VotesComponent API documentation].

Reviewer: > but I imagine we don't want to include all of this here right?
Right

    fn renounceRole(ref self: TState, role: felt252, account: ContractAddress);
}
----

== Votes

> The {votes-component} provides a flexible system for tracking voting power and delegation. This system allows users to delegate their voting power to other addresses, enabling more active participation in governance.

Reviewer: ```suggestion
The {votes-component} provides a flexible system for tracking voting power. This system allows users to delegate their voting power to other addresses, enabling more active participation in governance.
```

Author: Same as above. In the comment section of the component itself we are including "The Votes component provides a flexible system for tracking and delegating voting power..." and `delegate`and `delegate_by_sig`are two important functions of the component

Reviewer: The issue is not mentioning delegation, it is a matter of phrasing (@andrew-fleming can help here since I'm not a native speaker). To me "The Votes component provides a flexible system for tracking and delegating voting power" sounds ok, while "flexible system for tracking voting power and delegation" sounds weird, since the second sentence seems to imply that delegation, similar to voting power, is also tracked. Rephrasing the sentence could help, but just removing "and delegation" seemed ok, since that is also mentioned in the next sentence.

Reviewer 2: > To me "The Votes component provides a flexible system for tracking and delegating voting power" sounds ok, while "flexible system for tracking voting power and delegation" sounds weird, since the second sentence seems to imply that delegation, similar to voting power, is also tracked

Hmm yeah, I see what you mean. Even in a super obvious context like: "...provides a flexible system for eating pizza and delegation," the system obviously isn't for eating delegation, but IMO consistency with listed items offers a more clarified expression

Author: Went with "The {votes-component} provides a flexible system for tracking and delegating voting power. This system allows users to delegate their voting power to other addresses, enabling more active participation in governance." and the same for the above comment. Hope everyone is happy with that one.

NOTE: By default, token balance does not account for voting power. This makes transfers cheaper. The downside is that it requires users to delegate to themselves in order to activate checkpoints and have their voting power tracked.

> IMPORTANT: Transfering of voting units must be handled by the implementing contract. In the case of `ERC20` and `ERC721` this is usually done via the hooks. You can check the {votes-usage} section for examples on how to implement this.

Reviewer: ```suggestion
IMPORTANT: The transferring of voting units must be handled by the implementing contract. In the case of `ERC20` and `ERC721` this is usually done via the hooks. You can check the {votes-usage} section for examples of how to implement this.
```

=== Key Features

> 1. *Delegation*: Users can delegate their voting power to any address, including themselves.Vote power can be delegated either by calling

Reviewer: ```suggestion
1. *Delegation*: Users can delegate their voting power to any address, including themselves. Vote power can be delegated either by calling
```

the {delegate} function directly, or by providing a signature to be used with
{delegate_by_sig}.
2. *Historical lookups*: The system keeps track of voting power at different points in time, allowing for accurate voting in proposals that span multiple blocks.

> === Usage

Reviewer: ```suggestion
=== Usage

```

> When integrating the Votes component, the `VotingUnitsTrait` must be implemented to get the voting units for a given account as a function of the implementing contract. For simplicity, this module already provides two implementations for ERC20 and ERC721 tokens, which will work out of the box if the respective components are integrated.Additionally, you must embed the {src-component} to enable delegation by signatures.

Reviewer: ```suggestion
When integrating the Votes component, the `VotingUnitsTrait` must be implemented to get the voting units for a given account as a function of the implementing contract. For simplicity, this module already provides two implementations for ERC20 and ERC721 tokens, which will work out of the box if the respective components are integrated. Additionally, you must embed the {src-component} to enable delegation by signatures.
```

Reviewer: We try to split sentences into multiple lines with the main purpose of making them easier to follow here in reviews.

Reviewer: > Additionally, you must embed the {src-component} to enable delegation by signatures.

The {src-component} reference doesn't exist. Are we referring to NoncesComponent? If that's the case we should use "implement" instead of "embed" since we implement components and embed embeddable implementations.

Author: - Splitting into multilple lines: Good idea, makes it more readbale. But just FYI I don't think we've been very careful with that across the docs when inside the same paragraph.
- Updated the reference to nonce component and SNIP12Metadata
